the Chinook Database holds information about a music store. For the assignment, you will be assisting the Chinook team with understanding the media in their store, their customers and employees, and their invoice information by answering the questions below."
   

In [3]:
#importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


Question 1

which city has the best customers?we would like to throw a promotional Music Festival in the city we made the most money .Write a query that returns the 1 city that has the highest sum of invoice totals .Return both the city name and the sum of all invoices totals.


In [4]:
df = pd.read_csv('Invoice.csv')
df

InvoiceId  CustomerId  InvoiceDate  \
0            1           2          NaN   
1            2           4          NaN   
2            3           8          NaN   
3            4          14          NaN   
4            5          23          NaN   
..         ...         ...          ...   
407        408          25          NaN   
408        409          29          NaN   
409        410          35          NaN   
410        411          44          NaN   
411        412          58          NaN   

                               BillingAddress BillingCity BillingState  \
0                     Theodor-Heuss-Straße 34   Stuttgart          NaN   
1                            Ullevålsveien 14        Oslo          NaN   
2                             Grétrystraat 63    Brussels          NaN   
3                              8210 111 ST NW    Edmonton           AB   
4                             69 Salem Street      Boston           MA   
..                                        ...         ...          ...   
407                     319 N. Frances Street     Madison           WI   
408                    796 Dundas Street West     Toronto           ON   
409  Rua dos Campeões Europeus de Viena, 4350       Porto          NaN   
410                           Porthaninkatu 9    Helsinki          NaN   
411                       12,Community Centre       Delhi          NaN   

    BillingCountry BillingPostalCode  Total  
0          Germany             70174   1.98  
1           Norway              0171   3.96  
2          Belgium              1000   5.94  
3           Canada           T6G 2C7   8.91  
4              USA              2113  13.86  
..             ...               ...    ...  
407            USA             53703   3.96  
408         Canada           M6J 1V1   5.94  
409       Portugal               NaN   8.91  
410        Finland             00530  13.86  
411          India            110017   1.99  

[412 rows x 9 columns]

In [5]:
#To shows statistical summary of the data 
df.describe(include="all")

InvoiceId  CustomerId  InvoiceDate  BillingAddress BillingCity  \
count   412.000000  412.000000          0.0             412         412   
unique         NaN         NaN          NaN              59          53   
top            NaN         NaN          NaN  110 Raeburn Pl       Paris   
freq           NaN         NaN          NaN               7          14   
mean    206.500000   29.929612          NaN             NaN         NaN   
std     119.078406   17.010585          NaN             NaN         NaN   
min       1.000000    1.000000          NaN             NaN         NaN   
25%     103.750000   15.000000          NaN             NaN         NaN   
50%     206.500000   30.000000          NaN             NaN         NaN   
75%     309.250000   45.000000          NaN             NaN         NaN   
max     412.000000   59.000000          NaN             NaN         NaN   

       BillingState BillingCountry BillingPostalCode       Total  
count           210            412               384  412.000000  
unique           25             24                55         NaN  
top              CA            USA           EH4 1HH         NaN  
freq             21             91                 7         NaN  
mean            NaN            NaN               NaN    5.651942  
std             NaN            NaN               NaN    4.745320  
min             NaN            NaN               NaN    0.990000  
25%             NaN            NaN               NaN    1.980000  
50%             NaN            NaN               NaN    3.960000  
75%             NaN            NaN               NaN    8.910000  
max             NaN            NaN               NaN   25.860000

In [6]:
df.columns

Index(['InvoiceId', 'CustomerId', 'InvoiceDate', 'BillingAddress',
       'BillingCity', 'BillingState', 'BillingCountry', 'BillingPostalCode',
       'Total'],
      dtype='object')

In [7]:
bestcity_df =df[['BillingCity', 'Total']]
bestcity_df

BillingCity  Total
0     Stuttgart   1.98
1          Oslo   3.96
2      Brussels   5.94
3      Edmonton   8.91
4        Boston  13.86
..          ...    ...
407     Madison   3.96
408     Toronto   5.94
409       Porto   8.91
410    Helsinki  13.86
411       Delhi   1.99

[412 rows x 2 columns]

In [8]:
#to get the best city 

bestcity= bestcity_df.groupby("BillingCity")["Total"].sum().sort_values(ascending=False).head(1)
bestcity

BillingCity
Prague    90.24
Name: Total, dtype: float64

#the City with the best customer is Prague with the Total of 90.24 invoices

Question 2
which country has the most invoices?Use the invoice table to determine the countries that have the most invoices.Provide a table of BillingCountry and invoices ordered by the number of invoices for each country.The country with the most invoices should appear first.

In [9]:
#getting invoice dataframe
df

InvoiceId  CustomerId  InvoiceDate  \
0            1           2          NaN   
1            2           4          NaN   
2            3           8          NaN   
3            4          14          NaN   
4            5          23          NaN   
..         ...         ...          ...   
407        408          25          NaN   
408        409          29          NaN   
409        410          35          NaN   
410        411          44          NaN   
411        412          58          NaN   

                               BillingAddress BillingCity BillingState  \
0                     Theodor-Heuss-Straße 34   Stuttgart          NaN   
1                            Ullevålsveien 14        Oslo          NaN   
2                             Grétrystraat 63    Brussels          NaN   
3                              8210 111 ST NW    Edmonton           AB   
4                             69 Salem Street      Boston           MA   
..                                        ...         ...          ...   
407                     319 N. Frances Street     Madison           WI   
408                    796 Dundas Street West     Toronto           ON   
409  Rua dos Campeões Europeus de Viena, 4350       Porto          NaN   
410                           Porthaninkatu 9    Helsinki          NaN   
411                       12,Community Centre       Delhi          NaN   

    BillingCountry BillingPostalCode  Total  
0          Germany             70174   1.98  
1           Norway              0171   3.96  
2          Belgium              1000   5.94  
3           Canada           T6G 2C7   8.91  
4              USA              2113  13.86  
..             ...               ...    ...  
407            USA             53703   3.96  
408         Canada           M6J 1V1   5.94  
409       Portugal               NaN   8.91  
410        Finland             00530  13.86  
411          India            110017   1.99  

[412 rows x 9 columns]

In [10]:
bestcountry_df =df[['BillingCountry', 'InvoiceId']]
bestcountry_df

BillingCountry  InvoiceId
0          Germany          1
1           Norway          2
2          Belgium          3
3           Canada          4
4              USA          5
..             ...        ...
407            USA        408
408         Canada        409
409       Portugal        410
410        Finland        411
411          India        412

[412 rows x 2 columns]

In [11]:
#to get the BillingCountry with the highest invoiceId
bestcountry=bestcountry_df.groupby("BillingCountry")["InvoiceId"].count().sort_values(ascending=False).reset_index()
bestcountry

BillingCountry  InvoiceId
0              USA         91
1           Canada         56
2           Brazil         35
3           France         35
4          Germany         28
5   United Kingdom         21
6   Czech Republic         14
7         Portugal         14
8            India         13
9           Norway          7
10       Australia          7
11         Austria          7
12         Belgium          7
13          Sweden          7
14           Spain          7
15           Chile          7
16         Denmark          7
17     Netherlands          7
18         Finland          7
19          Poland          7
20         Hungary          7
21         Ireland          7
22           Italy          7
23       Argentina          7

The country with the highest invoice is USA with 91 total invoices

Question 3
Who is the best customer? The customer who has spend the most money will be decleared the best customer.Build a query that returns the person who has spend the most money.

In [12]:
df

InvoiceId  CustomerId  InvoiceDate  \
0            1           2          NaN   
1            2           4          NaN   
2            3           8          NaN   
3            4          14          NaN   
4            5          23          NaN   
..         ...         ...          ...   
407        408          25          NaN   
408        409          29          NaN   
409        410          35          NaN   
410        411          44          NaN   
411        412          58          NaN   

                               BillingAddress BillingCity BillingState  \
0                     Theodor-Heuss-Straße 34   Stuttgart          NaN   
1                            Ullevålsveien 14        Oslo          NaN   
2                             Grétrystraat 63    Brussels          NaN   
3                              8210 111 ST NW    Edmonton           AB   
4                             69 Salem Street      Boston           MA   
..                                        ...         ...          ...   
407                     319 N. Frances Street     Madison           WI   
408                    796 Dundas Street West     Toronto           ON   
409  Rua dos Campeões Europeus de Viena, 4350       Porto          NaN   
410                           Porthaninkatu 9    Helsinki          NaN   
411                       12,Community Centre       Delhi          NaN   

    BillingCountry BillingPostalCode  Total  
0          Germany             70174   1.98  
1           Norway              0171   3.96  
2          Belgium              1000   5.94  
3           Canada           T6G 2C7   8.91  
4              USA              2113  13.86  
..             ...               ...    ...  
407            USA             53703   3.96  
408         Canada           M6J 1V1   5.94  
409       Portugal               NaN   8.91  
410        Finland             00530  13.86  
411          India            110017   1.99  

[412 rows x 9 columns]

In [13]:
customer_df =pd.read_csv('Customer.csv')
customer_df

CustomerId  FirstName      LastName  \
0            1       Luís     Gonçalves   
1            2     Leonie        Köhler   
2            3   François      Tremblay   
3            4      Bjørn        Hansen   
4            5  František   Wichterlová   
5            6     Helena          Holý   
6            7     Astrid        Gruber   
7            8       Daan       Peeters   
8            9       Kara       Nielsen   
9           10    Eduardo       Martins   
10          11  Alexandre         Rocha   
11          12    Roberto       Almeida   
12          13   Fernanda         Ramos   
13          14       Mark       Philips   
14          15   Jennifer      Peterson   
15          16      Frank        Harris   
16          17       Jack         Smith   
17          18   Michelle        Brooks   
18          19        Tim         Goyer   
19          20        Dan        Miller   
20          21      Kathy         Chase   
21          22    Heather       Leacock   
22          23       John        Gordon   
23          24      Frank       Ralston   
24          25     Victor       Stevens   
25          26    Richard    Cunningham   
26          27    Patrick          Gray   
27          28      Julia       Barnett   
28          29     Robert         Brown   
29          30     Edward       Francis   
30          31     Martha          Silk   
31          32      Aaron      Mitchell   
32          33      Ellie      Sullivan   
33          34       João     Fernandes   
34          35   Madalena       Sampaio   
35          36     Hannah     Schneider   
36          37       Fynn    Zimmermann   
37          38     Niklas      Schröder   
38          39    Camille       Bernard   
39          40  Dominique      Lefebvre   
40          41       Marc        Dubois   
41          42      Wyatt        Girard   
42          43   Isabelle       Mercier   
43          44      Terhi    Hämäläinen   
44          45   Ladislav        Kovács   
45          46       Hugh      O'Reilly   
46          47      Lucas       Mancini   
47          48   Johannes  Van der Berg   
48          49  Stanisław        Wójcik   
49          50    Enrique         Muñoz   
50          51     Joakim     Johansson   
51          52       Emma         Jones   
52          53       Phil        Hughes   
53          54      Steve        Murray   
54          55       Mark        Taylor   
55          56      Diego     Gutiérrez   
56          57       Luis         Rojas   
57          58      Manoj        Pareek   
58          59       Puja    Srivastava   

                                             Company  \
0   Embraer - Empresa Brasileira de Aeronáutica S.A.   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                   JetBrains s.r.o.   
5                                                NaN   
6                                                NaN   
7                                                NaN   
8                                                NaN   
9                                   Woodstock Discos   
10                              Banco do Brasil S.A.   
11                                            Riotur   
12                                               NaN   
13                                             Telus   
14                                     Rogers Canada   
15                                       Google Inc.   
16                             Microsoft Corporation   
17                                               NaN   
18                                        Apple Inc.   
19                                               NaN   
20                                               NaN   
21                                               NaN   
22                                               NaN   
23                                               NaN   
24                     

In [14]:
#merge the invoice dataframe with the customer dataframe

customer = pd.merge(customer_df, df[['CustomerId', 'Total']], on = 'CustomerId', how = 'inner')
customer.head()

CustomerId FirstName   LastName  \
0           1      Luís  Gonçalves   
1           1      Luís  Gonçalves   
2           1      Luís  Gonçalves   
3           1      Luís  Gonçalves   
4           1      Luís  Gonçalves   

                                            Company  \
0  Embraer - Empresa Brasileira de Aeronáutica S.A.   
1  Embraer - Empresa Brasileira de Aeronáutica S.A.   
2  Embraer - Empresa Brasileira de Aeronáutica S.A.   
3  Embraer - Empresa Brasileira de Aeronáutica S.A.   
4  Embraer - Empresa Brasileira de Aeronáutica S.A.   

                           Address                 City State Country  \
0  Av. Brigadeiro Faria Lima, 2170  São José dos Campos    SP  Brazil   
1  Av. Brigadeiro Faria Lima, 2170  São José dos Campos    SP  Brazil   
2  Av. Brigadeiro Faria Lima, 2170  São José dos Campos    SP  Brazil   
3  Av. Brigadeiro Faria Lima, 2170  São José dos Campos    SP  Brazil   
4  Av. Brigadeiro Faria Lima, 2170  São José dos Campos    SP  Brazil   

  PostalCode               Phone                 Fax                 Email  \
0  12227-000  +55 (12) 3923-5555  +55 (12) 3923-5566  luisg@embraer.com.br   
1  12227-000  +55 (12) 3923-5555  +55 (12) 3923-5566  luisg@embraer.com.br   
2  12227-000  +55 (12) 3923-5555  +55 (12) 3923-5566  luisg@embraer.com.br   
3  12227-000  +55 (12) 3923-5555  +55 (12) 3923-5566  luisg@embraer.com.br   
4  12227-000  +55 (12) 3923-5555  +55 (12) 3923-5566  luisg@embraer.com.br   

   SupportRepId  Total  
0             3   3.98  
1             3   3.96  
2             3   5.94  
3             3   0.99  
4             3   1.98

In [15]:
#get the customer with highest bill 
customer.groupby('CustomerId')['Total'].sum().sort_values(ascending= False).head(1)

CustomerId
6    49.62
Name: Total, dtype: float64

customer 6 with a 49.62 total invoice



Question 4

Use your query to return the email,first name,last name and Genre of all Rock Music Listeners.Return your list ordered alphabetically by email address strating from A.


In [16]:
#get customer dataframe
customer_df

CustomerId  FirstName      LastName  \
0            1       Luís     Gonçalves   
1            2     Leonie        Köhler   
2            3   François      Tremblay   
3            4      Bjørn        Hansen   
4            5  František   Wichterlová   
5            6     Helena          Holý   
6            7     Astrid        Gruber   
7            8       Daan       Peeters   
8            9       Kara       Nielsen   
9           10    Eduardo       Martins   
10          11  Alexandre         Rocha   
11          12    Roberto       Almeida   
12          13   Fernanda         Ramos   
13          14       Mark       Philips   
14          15   Jennifer      Peterson   
15          16      Frank        Harris   
16          17       Jack         Smith   
17          18   Michelle        Brooks   
18          19        Tim         Goyer   
19          20        Dan        Miller   
20          21      Kathy         Chase   
21          22    Heather       Leacock   
22          23       John        Gordon   
23          24      Frank       Ralston   
24          25     Victor       Stevens   
25          26    Richard    Cunningham   
26          27    Patrick          Gray   
27          28      Julia       Barnett   
28          29     Robert         Brown   
29          30     Edward       Francis   
30          31     Martha          Silk   
31          32      Aaron      Mitchell   
32          33      Ellie      Sullivan   
33          34       João     Fernandes   
34          35   Madalena       Sampaio   
35          36     Hannah     Schneider   
36          37       Fynn    Zimmermann   
37          38     Niklas      Schröder   
38          39    Camille       Bernard   
39          40  Dominique      Lefebvre   
40          41       Marc        Dubois   
41          42      Wyatt        Girard   
42          43   Isabelle       Mercier   
43          44      Terhi    Hämäläinen   
44          45   Ladislav        Kovács   
45          46       Hugh      O'Reilly   
46          47      Lucas       Mancini   
47          48   Johannes  Van der Berg   
48          49  Stanisław        Wójcik   
49          50    Enrique         Muñoz   
50          51     Joakim     Johansson   
51          52       Emma         Jones   
52          53       Phil        Hughes   
53          54      Steve        Murray   
54          55       Mark        Taylor   
55          56      Diego     Gutiérrez   
56          57       Luis         Rojas   
57          58      Manoj        Pareek   
58          59       Puja    Srivastava   

                                             Company  \
0   Embraer - Empresa Brasileira de Aeronáutica S.A.   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                   JetBrains s.r.o.   
5                                                NaN   
6                                                NaN   
7                                                NaN   
8                                                NaN   
9                                   Woodstock Discos   
10                              Banco do Brasil S.A.   
11                                            Riotur   
12                                               NaN   
13                                             Telus   
14                                     Rogers Canada   
15                                       Google Inc.   
16                             Microsoft Corporation   
17                                               NaN   
18                                        Apple Inc.   
19                                               NaN   
20                                               NaN   
21                                               NaN   
22                                               NaN   
23                                               NaN   
24                     

In [17]:
#Import the Genre dataset
genre_df=pd.read_csv('Genre.csv')
genre_df

GenreId                Name
0         1                Rock
1         2                Jazz
2         3               Metal
3         4  Alternative & Punk
4         5       Rock And Roll
5         6               Blues
6         7               Latin
7         8              Reggae
8         9                 Pop
9        10          Soundtrack
10       11          Bossa Nova
11       12      Easy Listening
12       13         Heavy Metal
13       14            R&B/Soul
14       15   Electronica/Dance
15       16               World
16       17         Hip Hop/Rap
17       18     Science Fiction
18       19            TV Shows
19       20    Sci Fi & Fantasy
20       21               Drama
21       22              Comedy
22       23         Alternative
23       24           Classical
24       25               Opera

In [18]:
#Import the Track Data
track_df=pd.read_csv('Track.csv')
track_df

TrackId                                               Name  AlbumId  \
0           1            For Those About To Rock (We Salute You)        1   
1           2                                  Balls to the Wall        2   
2           3                                    Fast As a Shark        3   
3           4                                  Restless and Wild        3   
4           5                               Princess of the Dawn        3   
...       ...                                                ...      ...   
3498     3499  Pini Di Roma (Pinien Von Rom) \ I Pini Della V...      343   
3499     3500  String Quartet No. 12 in C Minor, D. 703 "Quar...      344   
3500     3501               L'orfeo, Act 3, Sinfonia (Orchestra)      345   
3501     3502  Quintet for Horn, Violin, 2 Violas, and Cello ...      346   
3502     3503                                      Koyaanisqatsi      347   

      MediaTypeId  GenreId                                           Composer  \
0               1        1          Angus Young, Malcolm Young, Brian Johnson   
1               2        1                                                NaN   
2               2        1  F. Baltes, S. Kaufman, U. Dirkscneider & W. Ho...   
3               2        1  F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. D...   
4               2        1                         Deaffy & R.A. Smith-Diesel   
...           ...      ...                                                ...   
3498            2       24                                                NaN   
3499            2       24                                     Franz Schubert   
3500            2       24                                 Claudio Monteverdi   
3501            2       24                            Wolfgang Amadeus Mozart   
3502            2       10                                       Philip Glass   

      Milliseconds     Bytes  UnitPrice  
0           343719  11170334       0.99  
1           342562   5510424       0.99  
2           230619   3990994       0.99  
3           252051   4331779       0.99  
4           375418   6290521       0.99  
...            ...       ...        ...  
3498        286741   4718950       0.99  
3499        139200   2283131       0.99  
3500         66639   1189062       0.99  
3501        221331   3665114       0.99  
3502        206005   3305164       0.99  

[3503 rows x 9 columns]

In [19]:
#Merge two data frames together
Gentrack=pd.merge(genre_df,track_df,how ="inner" ,on="GenreId")
Gentrack

GenreId     Name_x  TrackId  \
0           1       Rock        1   
1           1       Rock        2   
2           1       Rock        3   
3           1       Rock        4   
4           1       Rock        5   
...       ...        ...      ...   
3498       24  Classical     3499   
3499       24  Classical     3500   
3500       24  Classical     3501   
3501       24  Classical     3502   
3502       25      Opera     3451   

                                                 Name_y  AlbumId  MediaTypeId  \
0               For Those About To Rock (We Salute You)        1            1   
1                                     Balls to the Wall        2            2   
2                                       Fast As a Shark        3            2   
3                                     Restless and Wild        3            2   
4                                  Princess of the Dawn        3            2   
...                                                 ...      ...          ...   
3498  Pini Di Roma (Pinien Von Rom) \ I Pini Della V...      343            2   
3499  String Quartet No. 12 in C Minor, D. 703 "Quar...      344            2   
3500               L'orfeo, Act 3, Sinfonia (Orchestra)      345            2   
3501  Quintet for Horn, Violin, 2 Violas, and Cello ...      346            2   
3502  Die Zauberflöte, K.620: "Der Hölle Rache Kocht...      317            2   

                                               Composer  Milliseconds  \
0             Angus Young, Malcolm Young, Brian Johnson        343719   
1                                                   NaN        342562   
2     F. Baltes, S. Kaufman, U. Dirkscneider & W. Ho...        230619   
3     F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. D...        252051   
4                            Deaffy & R.A. Smith-Diesel        375418   
...                                                 ...           ...   
3498                                                NaN        286741   
3499                                     Franz Schubert        139200   
3500                                 Claudio Monteverdi         66639   
3501                            Wolfgang Amadeus Mozart        221331   
3502                            Wolfgang Amadeus Mozart        174813   

         Bytes  UnitPrice  
0     11170334       0.99  
1      5510424       0.99  
2      3990994       0.99  
3      4331779       0.99  
4      6290521       0.99  
...        ...        ...  
3498   4718950       0.99  
3499   2283131       0.99  
3500   1189062       0.99  
3501   3665114       0.99  
3502   2861468       0.99  

[3503 rows x 10 columns]

In [20]:
Genre=Gentrack[['GenreId', 'Name_x', 'TrackId',]]
Genre

GenreId     Name_x  TrackId
0           1       Rock        1
1           1       Rock        2
2           1       Rock        3
3           1       Rock        4
4           1       Rock        5
...       ...        ...      ...
3498       24  Classical     3499
3499       24  Classical     3500
3500       24  Classical     3501
3501       24  Classical     3502
3502       25      Opera     3451

[3503 rows x 3 columns]

In [21]:
RockGenre=Genre.groupby(["Name_x"]).count()
RockGenre

GenreId  TrackId
Name_x                              
Alternative              40       40
Alternative & Punk      332      332
Blues                    81       81
Bossa Nova               15       15
Classical                74       74
Comedy                   17       17
Drama                    64       64
Easy Listening           24       24
Electronica/Dance        30       30
Heavy Metal              28       28
Hip Hop/Rap              35       35
Jazz                    130      130
Latin                   579      579
Metal                   374      374
Opera                     1        1
Pop                      48       48
R&B/Soul                 61       61
Reggae                   58       58
Rock                   1297     1297
Rock And Roll            12       12
Sci Fi & Fantasy         26       26
Science Fiction          13       13
Soundtrack               43       43
TV Shows                 93       93
World                    28       28

In [22]:
#Import the InvoiceLine dataset
df4=pd.read_csv('InvoiceLine.csv')
df4

InvoiceLineId  InvoiceId  TrackId  UnitPrice  Quantity
0                 1          1        2       0.99         1
1                 2          1        4       0.99         1
2                 3          2        6       0.99         1
3                 4          2        8       0.99         1
4                 5          2       10       0.99         1
...             ...        ...      ...        ...       ...
2235           2236        411     3136       0.99         1
2236           2237        411     3145       0.99         1
2237           2238        411     3154       0.99         1
2238           2239        411     3163       0.99         1
2239           2240        412     3177       1.99         1

[2240 rows x 5 columns]

In [23]:
query=customer_df[['Email', 'FirstName', 'LastName',]]
query

Email  FirstName      LastName
0            luisg@embraer.com.br       Luís     Gonçalves
1           leonekohler@surfeu.de     Leonie        Köhler
2             ftremblay@gmail.com   François      Tremblay
3           bjorn.hansen@yahoo.no      Bjørn        Hansen
4        frantisekw@jetbrains.com  František   Wichterlová
5                 hholy@gmail.com     Helena          Holý
6          astrid.gruber@apple.at     Astrid        Gruber
7           daan_peeters@apple.be       Daan       Peeters
8           kara.nielsen@jubii.dk       Kara       Nielsen
9        eduardo@woodstock.com.br    Eduardo       Martins
10               alero@uol.com.br  Alexandre         Rocha
11  roberto.almeida@riotur.gov.br    Roberto       Almeida
12       fernadaramos4@uol.com.br   Fernanda         Ramos
13             mphilips12@shaw.ca       Mark       Philips
14            jenniferp@rogers.ca   Jennifer      Peterson
15             fharris@google.com      Frank        Harris
16        jacksmith@microsoft.com       Jack         Smith
17              michelleb@aol.com   Michelle        Brooks
18               tgoyer@apple.com        Tim         Goyer
19            dmiller@comcast.com        Dan        Miller
20            kachase@hotmail.com      Kathy         Chase
21             hleacock@gmail.com    Heather       Leacock
22         johngordon22@yahoo.com       John        Gordon
23             fralston@gmail.com      Frank       Ralston
24             vstevens@yahoo.com     Victor       Stevens
25       ricunningham@hotmail.com    Richard    Cunningham
26           patrick.gray@aol.com    Patrick          Gray
27            jubarnett@gmail.com      Julia       Barnett
28               robbrown@shaw.ca     Robert         Brown
29            edfrancis@yachoo.ca     Edward       Francis
30           marthasilk@gmail.com     Martha          Silk
31         aaronmitchell@yahoo.ca      Aaron      Mitchell
32         ellie.sullivan@shaw.ca      Ellie      Sullivan
33            jfernandes@yahoo.pt       João     Fernandes
34              masampaio@sapo.pt   Madalena       Sampaio
35      hannah.schneider@yahoo.de     Hannah     Schneider
36           fzimmermann@yahoo.de       Fynn    Zimmermann
37            nschroder@surfeu.de     Niklas      Schröder
38       camille.bernard@yahoo.fr    Camille       Bernard
39    dominiquelefebvre@gmail.com  Dominique      Lefebvre
40        marc.dubois@hotmail.com       Marc        Dubois
41          wyatt.girard@yahoo.fr      Wyatt        Girard
42      isabelle_mercier@apple.fr   Isabelle       Mercier
43      terhi.hamalainen@apple.fi      Terhi    Hämäläinen
44       ladislav_kovacs@apple.hu   Ladislav        Kovács
45           hughoreilly@apple.ie       Hugh      O'Reilly
46         lucas.mancini@yahoo.it      Lucas       Mancini
47        johavanderberg@yahoo.nl   Johannes  Van der Berg
48         stanisław.wójcik@wp.pl  Stanisław        Wójcik
49         enrique_munoz@yahoo.es    Enrique         Muñoz
50      joakim.johansson@yahoo.se     Joakim     Johansson
51         emma_jones@hotmail.com       Emma         Jones
52          phil.hughes@gmail.com       Phil        Hughes
53          steve.murray@yahoo.uk      Steve        Murray
54           mark.taylor@yahoo.au       Mark        Taylor
55       diego.gutierrez@yahoo.ar      Diego     Gutiérrez
56             luisrojas@yahoo.cl       Luis         Rojas
57        manoj.pareek@rediff.com      Manoj        Pareek
58       puja_srivastava@yahoo.in       Puja    Srivastava

In [24]:
#Sort values Alphabetically
query.sort_values(["Email"], axis = 0, ascending = True)

Email  FirstName      LastName
31         aaronmitchell@yahoo.ca      Aaron      Mitchell
10               alero@uol.com.br  Alexandre         Rocha
6          astrid.gruber@apple.at     Astrid        Gruber
3           bjorn.hansen@yahoo.no      Bjørn        Hansen
38       camille.bernard@yahoo.fr    Camille       Bernard
7           daan_peeters@apple.be       Daan       Peeters
55       diego.gutierrez@yahoo.ar      Diego     Gutiérrez
19            dmiller@comcast.com        Dan        Miller
39    dominiquelefebvre@gmail.com  Dominique      Lefebvre
29            edfrancis@yachoo.ca     Edward       Francis
9        eduardo@woodstock.com.br    Eduardo       Martins
32         ellie.sullivan@shaw.ca      Ellie      Sullivan
51         emma_jones@hotmail.com       Emma         Jones
49         enrique_munoz@yahoo.es    Enrique         Muñoz
12       fernadaramos4@uol.com.br   Fernanda         Ramos
15             fharris@google.com      Frank        Harris
23             fralston@gmail.com      Frank       Ralston
4        frantisekw@jetbrains.com  František   Wichterlová
2             ftremblay@gmail.com   François      Tremblay
36           fzimmermann@yahoo.de       Fynn    Zimmermann
35      hannah.schneider@yahoo.de     Hannah     Schneider
5                 hholy@gmail.com     Helena          Holý
21             hleacock@gmail.com    Heather       Leacock
45           hughoreilly@apple.ie       Hugh      O'Reilly
42      isabelle_mercier@apple.fr   Isabelle       Mercier
16        jacksmith@microsoft.com       Jack         Smith
14            jenniferp@rogers.ca   Jennifer      Peterson
33            jfernandes@yahoo.pt       João     Fernandes
50      joakim.johansson@yahoo.se     Joakim     Johansson
47        johavanderberg@yahoo.nl   Johannes  Van der Berg
22         johngordon22@yahoo.com       John        Gordon
27            jubarnett@gmail.com      Julia       Barnett
20            kachase@hotmail.com      Kathy         Chase
8           kara.nielsen@jubii.dk       Kara       Nielsen
44       ladislav_kovacs@apple.hu   Ladislav        Kovács
1           leonekohler@surfeu.de     Leonie        Köhler
46         lucas.mancini@yahoo.it      Lucas       Mancini
0            luisg@embraer.com.br       Luís     Gonçalves
56             luisrojas@yahoo.cl       Luis         Rojas
57        manoj.pareek@rediff.com      Manoj        Pareek
40        marc.dubois@hotmail.com       Marc        Dubois
54           mark.taylor@yahoo.au       Mark        Taylor
30           marthasilk@gmail.com     Martha          Silk
34              masampaio@sapo.pt   Madalena       Sampaio
17              michelleb@aol.com   Michelle        Brooks
13             mphilips12@shaw.ca       Mark       Philips
37            nschroder@surfeu.de     Niklas      Schröder
26           patrick.gray@aol.com    Patrick          Gray
52          phil.hughes@gmail.com       Phil        Hughes
58       puja_srivastava@yahoo.in       Puja    Srivastava
25       ricunningham@hotmail.com    Richard    Cunningham
28               robbrown@shaw.ca     Robert         Brown
11  roberto.almeida@riotur.gov.br    Roberto       Almeida
48         stanisław.wójcik@wp.pl  Stanisław        Wójcik
53          steve.murray@yahoo.uk      Steve        Murray
43      terhi.hamalainen@apple.fi      Terhi    Hämäläinen
18               tgoyer@apple.com        Tim         Goyer
24             vstevens@yahoo.com     Victor       Stevens
41          wyatt.girard@yahoo.fr      Wyatt        Girard

In [25]:
#Sort values Alphabetically
query.sort_values(["Email"], axis = 0, ascending = True)

Email  FirstName      LastName
31         aaronmitchell@yahoo.ca      Aaron      Mitchell
10               alero@uol.com.br  Alexandre         Rocha
6          astrid.gruber@apple.at     Astrid        Gruber
3           bjorn.hansen@yahoo.no      Bjørn        Hansen
38       camille.bernard@yahoo.fr    Camille       Bernard
7           daan_peeters@apple.be       Daan       Peeters
55       diego.gutierrez@yahoo.ar      Diego     Gutiérrez
19            dmiller@comcast.com        Dan        Miller
39    dominiquelefebvre@gmail.com  Dominique      Lefebvre
29            edfrancis@yachoo.ca     Edward       Francis
9        eduardo@woodstock.com.br    Eduardo       Martins
32         ellie.sullivan@shaw.ca      Ellie      Sullivan
51         emma_jones@hotmail.com       Emma         Jones
49         enrique_munoz@yahoo.es    Enrique         Muñoz
12       fernadaramos4@uol.com.br   Fernanda         Ramos
15             fharris@google.com      Frank        Harris
23             fralston@gmail.com      Frank       Ralston
4        frantisekw@jetbrains.com  František   Wichterlová
2             ftremblay@gmail.com   François      Tremblay
36           fzimmermann@yahoo.de       Fynn    Zimmermann
35      hannah.schneider@yahoo.de     Hannah     Schneider
5                 hholy@gmail.com     Helena          Holý
21             hleacock@gmail.com    Heather       Leacock
45           hughoreilly@apple.ie       Hugh      O'Reilly
42      isabelle_mercier@apple.fr   Isabelle       Mercier
16        jacksmith@microsoft.com       Jack         Smith
14            jenniferp@rogers.ca   Jennifer      Peterson
33            jfernandes@yahoo.pt       João     Fernandes
50      joakim.johansson@yahoo.se     Joakim     Johansson
47        johavanderberg@yahoo.nl   Johannes  Van der Berg
22         johngordon22@yahoo.com       John        Gordon
27            jubarnett@gmail.com      Julia       Barnett
20            kachase@hotmail.com      Kathy         Chase
8           kara.nielsen@jubii.dk       Kara       Nielsen
44       ladislav_kovacs@apple.hu   Ladislav        Kovács
1           leonekohler@surfeu.de     Leonie        Köhler
46         lucas.mancini@yahoo.it      Lucas       Mancini
0            luisg@embraer.com.br       Luís     Gonçalves
56             luisrojas@yahoo.cl       Luis         Rojas
57        manoj.pareek@rediff.com      Manoj        Pareek
40        marc.dubois@hotmail.com       Marc        Dubois
54           mark.taylor@yahoo.au       Mark        Taylor
30           marthasilk@gmail.com     Martha          Silk
34              masampaio@sapo.pt   Madalena       Sampaio
17              michelleb@aol.com   Michelle        Brooks
13             mphilips12@shaw.ca       Mark       Philips
37            nschroder@surfeu.de     Niklas      Schröder
26           patrick.gray@aol.com    Patrick          Gray
52          phil.hughes@gmail.com       Phil        Hughes
58       puja_srivastava@yahoo.in       Puja    Srivastava
25       ricunningham@hotmail.com    Richard    Cunningham
28               robbrown@shaw.ca     Robert         Brown
11  roberto.almeida@riotur.gov.br    Roberto       Almeida
48         stanisław.wójcik@wp.pl  Stanisław        Wójcik
53          steve.murray@yahoo.uk      Steve        Murray
43      terhi.hamalainen@apple.fi      Terhi    Hämäläinen
18               tgoyer@apple.com        Tim         Goyer
24             vstevens@yahoo.com     Victor       Stevens
41          wyatt.girard@yahoo.fr      Wyatt        Girard


Question 5

We want to find out the most popular music genre for each country.we determine the most popular genre as the highest amount of purchases.Write a query that returns each country along with the top genre with the highest amount of purchases.Write a query that returns each country along with the top genre for countries where the maximum number of purchases is shared all genres.For this query,you will need to use the Invoice,InvoiceLine,Customer,Track,and Genre tables.


In [27]:
#import the invoice line ID dataframe
df4

InvoiceLineId  InvoiceId  TrackId  UnitPrice  Quantity
0                 1          1        2       0.99         1
1                 2          1        4       0.99         1
2                 3          2        6       0.99         1
3                 4          2        8       0.99         1
4                 5          2       10       0.99         1
...             ...        ...      ...        ...       ...
2235           2236        411     3136       0.99         1
2236           2237        411     3145       0.99         1
2237           2238        411     3154       0.99         1
2238           2239        411     3163       0.99         1
2239           2240        412     3177       1.99         1

[2240 rows x 5 columns]

In [28]:
Customer_ = pd.merge(df4,df[['InvoiceId', 'CustomerId', 'BillingCountry']], on ='InvoiceId', how = 'left')
Customer_

InvoiceLineId  InvoiceId  TrackId  UnitPrice  Quantity  CustomerId  \
0                 1          1        2       0.99         1           2   
1                 2          1        4       0.99         1           2   
2                 3          2        6       0.99         1           4   
3                 4          2        8       0.99         1           4   
4                 5          2       10       0.99         1           4   
...             ...        ...      ...        ...       ...         ...   
2235           2236        411     3136       0.99         1          44   
2236           2237        411     3145       0.99         1          44   
2237           2238        411     3154       0.99         1          44   
2238           2239        411     3163       0.99         1          44   
2239           2240        412     3177       1.99         1          58   

     BillingCountry  
0           Germany  
1           Germany  
2            Norway  
3            Norway  
4            Norway  
...             ...  
2235        Finland  
2236        Finland  
2237        Finland  
2238        Finland  
2239          India  

[2240 rows x 7 columns]

In [29]:
#merge genre dataframe with customer_
Customer_ = pd.merge(Customer_,Genre,on = 'TrackId',how ='left')
Customer_

InvoiceLineId  InvoiceId  TrackId  UnitPrice  Quantity  CustomerId  \
0                 1          1        2       0.99         1           2   
1                 2          1        4       0.99         1           2   
2                 3          2        6       0.99         1           4   
3                 4          2        8       0.99         1           4   
4                 5          2       10       0.99         1           4   
...             ...        ...      ...        ...       ...         ...   
2235           2236        411     3136       0.99         1          44   
2236           2237        411     3145       0.99         1          44   
2237           2238        411     3154       0.99         1          44   
2238           2239        411     3163       0.99         1          44   
2239           2240        412     3177       1.99         1          58   

     BillingCountry  GenreId    Name_x  
0           Germany        1      Rock  
1           Germany        1      Rock  
2            Norway        1      Rock  
3            Norway        1      Rock  
4            Norway        1      Rock  
...             ...      ...       ...  
2235        Finland        3     Metal  
2236        Finland        3     Metal  
2237        Finland        7     Latin  
2238        Finland        7     Latin  
2239          India       19  TV Shows  

[2240 rows x 9 columns]

In [30]:
#get total price for each country by genre
country_total =pd.DataFrame(Customer_.groupby(['BillingCountry','Name_x'], as_index=False)['UnitPrice'].sum()) 
country_total

BillingCountry              Name_x  UnitPrice
0         Argentina  Alternative & Punk       8.91
1         Argentina      Easy Listening       1.98
2         Argentina                Jazz       1.98
3         Argentina               Latin       7.92
4         Argentina               Metal       6.93
..              ...                 ...        ...
232  United Kingdom                 Pop       1.98
233  United Kingdom            R&B/Soul       1.98
234  United Kingdom              Reggae       4.95
235  United Kingdom                Rock      36.63
236  United Kingdom               World       0.99

[237 rows x 3 columns]

In [31]:
#max total for each country
total =pd.DataFrame(country_total.groupby('BillingCountry', as_index=False)['UnitPrice'].max())
total

BillingCountry  UnitPrice
0        Argentina       8.91
1        Australia      21.78
2          Austria      14.85
3          Belgium      20.79
4           Brazil      80.19
5           Canada     105.93
6            Chile       8.91
7   Czech Republic      24.75
8          Denmark      20.79
9          Finland      17.82
10          France      64.35
11         Germany      61.38
12         Hungary      10.89
13           India      24.75
14         Ireland      13.93
15           Italy      17.82
16     Netherlands      17.82
17          Norway      16.83
18          Poland      21.78
19        Portugal      30.69
20           Spain      21.78
21          Sweden      11.88
22             USA     155.43
23  United Kingdom      36.63

In [32]:
#highest genre by country
pd.merge(total,country_total, on =['UnitPrice','BillingCountry'], how ='left')

BillingCountry  UnitPrice              Name_x
0        Argentina       8.91  Alternative & Punk
1        Argentina       8.91                Rock
2        Australia      21.78                Rock
3          Austria      14.85                Rock
4          Belgium      20.79                Rock
5           Brazil      80.19                Rock
6           Canada     105.93                Rock
7            Chile       8.91                Rock
8   Czech Republic      24.75                Rock
9          Denmark      20.79                Rock
10         Finland      17.82                Rock
11          France      64.35                Rock
12         Germany      61.38                Rock
13         Hungary      10.89                Rock
14           India      24.75                Rock
15         Ireland      13.93            TV Shows
16           Italy      17.82                Rock
17     Netherlands      17.82                Rock
18          Norway      16.83                Rock
19          Poland      21.78                Rock
20        Portugal      30.69                Rock
21           Spain      21.78                Rock
22          Sweden      11.88               Latin
23             USA     155.43                Rock
24  United Kingdom      36.63                Rock